### ISSUE TITLE
Get all option values via `rnix-parser` and `rnix-lsp`

### ISSUE THREAD

**Issue by lapp0** [lapp0/nix-gui#9]
# Currently

Default values are retrieved via
```
        subprocess.run([
            'nix-build',
            '/etc/nixos/nixpkgs/nixos/release.nix',
            '-A',
            'options',
            '-o',
            release_path
        ])
```

User set values are retrieved from `CONFIGURATION_PATH` via `unsafeGetAttrPos` and an interface to `rnix-parser`

# Desired

- `parseoptions.rs` (based on `rnix-parser`) is used to parse all of nixpkgs and `CONFIGURATION_PATH`
- `parseoptions.rs` spits out `module`, `option`, `value_expression`
- `api.py` calls `parseoptions.rs` and retrieves all option values

**Comment by aaronjanse** [comment]
Hmm, I don't think it would be possible to get option values via a parser alone. For example, if someone writes `services.foo = 1 + 2`, you would need to know that `1 + 2 == 3`.

Maybe you could use `nix-instantiate --eval` to output JSON with the values of all NixOS options? I could help you write the `nix-instantiate` command to do this :-)

**Comment by lapp0** [comment]
@aaronjanse good idea. After discussing with @adrianparvin it seems like the best route to go involves `nix-instantiate --eval` rather than parsing to find the options. To retrieve the unevaluated "default" expression for each option the best route may be
- 1) Create an expression that retrieves the module, line, and column of each option within `<nixpkgs/nixos>`
- 2) For each module, extract its ast
- 3) locate the option expression in the same way it's done currently `configuration.nix` https://github.com/lapp0/nix-gui/blob/master/nixui/parser/parser.py#L10-L25

This would be performed each time `<nixpkgs/nixos>` is changed.

Any thoughts on going about getting default expressions this way?

**Comment by aaronjanse** [comment]
> To retrieve the unevaluated "default" expression for each option

Are you referring to Nixpkgs's default values, or are you referring to the values that users put in their NixOS config before opening nix-gui?

**Comment by lapp0** [comment]
I'm referring to values nixpkgs default values. Currently they're retrieved by building `release.nix` and caching it for later runs of the program https://github.com/lapp0/nix-gui/blob/master/nixui/api.py#L22-L32

**Comment by aaronjanse** [comment]
Ah got it. Is there a reason to not just keep using the `nix-build` expression described in the issue description?

**Comment by lapp0** [comment]
1) Building `release.nix` is somewhat slow
2) More importantly, we cannot get the unevaluated form of an expression through building, but we can via `unsafeGetAttrPos` + `dump-ast`. This is helpful when the user wants to make an expression based off of the default expression, and in cases where the default expression is complex or uses string substitution.

**Comment by adrianparvino** [comment]
I opened #20 and #24 in order to identify the problems in the current implementation, and to avoid the pitfalls in any future solution.

My initial identification of the problem(quoted from Matrix):

> So #9's problems are split into 2 parts, right?
> (a) The default values set by NixOS itself.
> And (b) the values (and locations of their definitions) by the user.
> (a) is kinda mostly solved, but it could be done better.
> (b) needs a less intensive implementation, and that's where `rnix-parser` comes into play.

@lapp0's response was:

> yes, that's mostly the picture
> (a)'s current solution is lacking: it doesn't allow extraction of expressions, rather it just retrieves the evaluated expression, 
> however for (b), the expression is extracted then evaluated so either can be edited.
> For (b), rnix-parser is already used to extract the AST. To actually figure out where an option is defined we must evaluate (I think?), and I don't think it'd be wise to reimplement nix evaluation using rnix-parser for this project.
> Therefore, it probably makes sense to:
> 1. keep track of the state of nixpkgs
> 2. load and disk-cache the location (module, line, column) of each option with a recursive expression utilizing unsafeGetAttrPosn similar to what we have now, but in pure nix rather than recursion implemented through python (in my testing, 100 nix-instantiate calls to get 100 attribute positions is about the same cost as 1 call getting 100 attribute positions)
> 3. at runtime, lookup and disk-cache ASTs in order to recover the expression string

I have not yet provided a proper solution for (a). My recent PRs however made me realize that (a) and (b) can (and should be) solved at the same time.

While implementing #24, I realized a few important things. The results of (a) depends on (b):

The current implementation assumes that all definitions are attrsets, when it is not true for submodules, and with the new `extraConfig` → `settings` change (https://github.com/NixOS/rfcs/pull/42), this is even further to be expected.

There are 5 axes in a NixOS module:

1. Is the module a function(from attrset to attrset) or an attrset?
2. Does the module have an `imports`?
3. Does the module have an `options`?
4. Does the module have a `config`?
5. Is the module in canonical form?

Currently, our parser and extractor only accepts modules that are a function, in non-canonical form. That is:

```nix
{ ... }:
{ 
  ...
}
```
such that the result attrset is the `config` but optionally contains `imports`.

This is enough to implement most `configuration.nix` however, most `<nixpkgs/nixos>` modules are in the canonical form. To make it even more difficult, `config` is often not implemented directly as an attrset(`{ ... config = { ... };  }` however they are implemented as (`{ ... config = mkIf ... { ... }}`), as most `<nixpkgs/nixos>` modules assume the existence of a module system. Traversing them is more difficult due to this.

Furthermore, an option can implement `merge`, that is, there can be multiple `config` definitions for them. This fact probably deserves a separate issue.

`imports` is also its own separate problem, as it can contain full NixOS modules aside from file paths to NixOS modules.

My current proposal is to use the NixOS module system in order to detect the declaration of options. This is available under `options` of `<nixpkgs/nixos>`. I recently had to upgrade a 20.09 laptop to 21.05 and those options proved themselves to be useful in order to track down how an object is defined. Multiple passes can be done, the evaluation pass and the parsing pass. With the evaluation pass, the configuration is built against NixOS, this allows us to obtain the `options`. The `config` doesn't seem to be currently important. With `options` obtained, we are able to obtain an important information: its path. We are then able to make a closure of all option paths. With that obtained, the parsing pass starts, we canonicalize it by detecting if the attrset contains the `options` key. We are able to obtain a `default` expression mapping by simply traversing over the `options` attribute, and the definition expressions by simply traversing over the `config` attribute, with the caveat of it being aware of `mkIf ...`. We then recursively do this over `imports` which are Nix expressions aside from file paths.

CC: @lapp0 @aaronjanse

**Comment by lapp0** [comment]
Thanks for the writeup @adrianparvino 

>This is enough to implement most configuration.nix however, most <nixpkgs/nixos> modules are in the canonical form. To make it even more difficult, config is often not implemented directly as an attrset({ ... config = { ... }; } however they are implemented as ({ ... config = mkIf ... { ... }}), as most <nixpkgs/nixos> modules assume the existence of a module system. Traversing them is more difficult due to this.

How does `unsafeGetAttrPos` behave for `mkIf`? For example, if `config.programs.atop.settings != { }` does it return this position? https://github.com/NixOS/nixpkgs/blob/ab51a2dbd6b5fcc8f6ae467075a2863ecac7c7d0/nixos/modules/programs/atop.nix#L115-L120

I'm probably not quite understanding how `config` interacts with our setup here. Is part of the problem that the dynamic nature of `config` make it difficult to determine whether `config.programs.atop.settings != { }` in the preprocessing step? I'm not quite following why we'd determine `cfg` **after** obtaining the `options`.

I'd be curious to hear Aarons thoughts on your proposal.

**Comment by aaronjanse** [comment]
Is the goal to replicate the information shown in NixOS option search \[1], which shows text like `pkgs.minecraft-server` for default values (rather than showing `<<derivation>>`)?

If so, I might be able to help write a faster version of the release.nix thing for this use-case

[1] https://search.nixos.org/options

**Comment by lapp0** [comment]
Not quite. For something like `dysnomia.package`, there is no `defaultText` set, and the expression is 
```
dysnomia.package = pkgs.dysnomia.override (origArgs: dysnomiaFlags // lib.optionalAttrs (cfg.enableLegacyModules) {
      enableLegacy = builtins.trace ''
        WARNING: Dysnomia has been configured to use the legacy 'process' and 'wrapper'
        modules for compatibility reasons! If you rely on these modules, consider
        migrating to better alternatives.
        More information: https://raw.githubusercontent.com/svanderburg/dysnomia/f65a9a84827bcc4024d6b16527098b33b02e4054/README-legacy.md
        If you have migrated already or don't rely on these Dysnomia modules, you can
        disable legacy mode with the following NixOS configuration option:
        dysnomia.enableLegacyModules = false;
        In a future version of Dysnomia (and NixOS) the legacy option will go away!
      '' true;
    });
```
So using the same method as https://search.nixos.org/options wouldn't allow us to extract the actual default expression.

**Here's what I'm thinking the ideal situation would look like, let me know if I'm on the right track here:**

Users are able to edit expressions using either a type-appropriate form (e.g. for `zramSwap.algorithm` they can choose between a radio button group for "one of" or a textbox for setting a string)
![image](https://user-images.githubusercontent.com/2560255/124519844-96e14a00-ddb8-11eb-8d0d-dec3100e3fa6.png)

The evaluated value of the expression will be shown by default, but users will be able to edit it multiple ways. For example let's say `zramSwap.algorithm = if true then "lz4" else "lza";`. In this case, users will be able to edit the option with three different forms:
- "one of" radio button group (working)
- "string" text box (working)
- "expression" text box (not working until after this issue is resolved)

![image](https://user-images.githubusercontent.com/2560255/124520424-77e3b780-ddba-11eb-8ffa-844ca5aaced0.png)

![image](https://user-images.githubusercontent.com/2560255/124520473-977ae000-ddba-11eb-9bcc-b9e2c94415d9.png)

(mockup)

![image](https://user-images.githubusercontent.com/2560255/124520862-b9289700-ddbb-11eb-8ccb-812a943f122d.png)

**Comment by adrianparvino** [comment]
Current extraction code fails for the following files:

```python
['nixos/modules/i18n/input-method/kime.nix', 'nixos/modules/misc/documentation.nix', 'nixos/modules/misc/nixops-autoluks.nix', 'nixos/modules/misc/nixpkgs.nix', 'nixos/modules/programs/oblogout.nix', 'nixos/modules/rename.nix', 'nixos/modules/security/acme.nix', 'nixos/modules/security/rngd.nix', 'nixos/modules/services/cluster/kubernetes/controller-manager.nix', 'nixos/modules/services/cluster/kubernetes/kubelet.nix', 'nixos/modules/services/cluster/kubernetes/proxy.nix', 'nixos/modules/services/cluster/kubernetes/scheduler.nix', 'nixos/modules/services/continuous-integration/hercules-ci-agent/common.nix', 'nixos/modules/services/continuous-integration/hercules-ci-agent/default.nix', 'nixos/modules/services/databases/neo4j.nix', 'nixos/modules/services/desktops/pantheon/files.nix', 'nixos/modules/services/mail/dovecot.nix', 'nixos/modules/services/mail/rspamd.nix', 'nixos/modules/services/monitoring/grafana.nix', 'nixos/modules/services/monitoring/prometheus/exporters.nix', 'nixos/modules/services/network-filesystems/ipfs.nix', 'nixos/modules/services/networking/biboumi.nix', 'nixos/modules/services/networking/bird.nix', 'nixos/modules/services/networking/matterbridge.nix', 'nixos/modules/services/networking/ntp/openntpd.nix', 'nixos/modules/services/networking/pleroma.nix', 'nixos/modules/services/networking/searx.nix', 'nixos/modules/services/system/dbus.nix', 'nixos/modules/services/torrent/transmission.nix', 'nixos/modules/services/web-apps/discourse.nix', 'nixos/modules/services/x11/display-managers/slim.nix', 'nixos/modules/system/activation/top-level.nix', 'nixos/modules/tasks/network-interfaces.nix', 'nixos/modules/virtualisation/oci-containers.nix', 'nixos/modules/virtualisation/podman-network-socket-ghostunnel.nix', 'nixos/modules/virtualisation/podman-network-socket.nix']
```
